### PR TITLE
Variable Design.

### PR DESCRIPTION
**Pull request by qingqing01**
This design was completed with @hedaoyuan, @pkuyym  @reyoung , @QiJune and @Canpio .

Fix https://github.com/PaddlePaddle/Paddle/issues/2545

### CODE REVIEW

File: paddle/framework/variable.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+  // If the value is nullptr, create a new object, then return it. And
+  // the type T should have a default constructor.
+  // The `is_new_obj` can use be used to determine whether the returned pointer
+  // is a new object.
+  template <class T>
+  T* GetMutable(bool* is_new_obj=nullptr);
+
+privated:
+  // A typed data pointer, set the type when calling Reset.
+  boost::any value_;
```
Comment (wangkuiyi) [inline paddle/framework/variable.md]: I guess we are going to use boost::any for both saving the pointer and the type information (caffe2::TypeMeta)? 

I know that boost::any has a method type(), which returns a typeid comparable with typeid(TypeName).  But how could we connect the typeid with a deleter?

File: paddle/framework/variable.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+* Accept external data pointer of any type
+* Get typed data
+* Destroy data in destructor
+
+## Prototype
+
+Two aspects should be considered when designing `Variable`, the first is interactive interfaces with external data, the second is  type maintaining.  Here, we use `boost::any` to hold data and its type.
+
+```cpp
+Class Variable {
```
Comment (wangkuiyi) [inline paddle/framework/variable.md]: Class ==> class

File: paddle/framework/variable.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  template <class T>
+  T* GetMutable(bool* is_new_obj=nullptr);
+
+privated:
+  // A typed data pointer, set the type when calling Reset.
+  boost::any value_;
+  
+  // A destroy call.
+  template <class T>
+  static void Destroy(void* pointer);
```
Comment (wangkuiyi) [inline paddle/framework/variable.md]: In C++, there are two concepts related with here -- destruction and delete.  Let us not invent a name destroy, but just call it deleter, as [in shared_ptr::shared_ptr](http://en.cppreference.com/w/cpp/memory/shared_ptr/shared_ptr).

File: paddle/framework/variable.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  
+  // A pointer to save the destructor.
+  typedef void (*DestroyCall)(void *);
```
Comment (wangkuiyi) [inline paddle/framework/variable.md]: As we represent value_ by boost::any, I think we are not going to use void* in type DestroyCall?

How about 

```cpp
std::function<void(boost::any&)> deleter_;
```

**Review by wangkuiyi — APPROVED** [review]
I understand that we want to use boost::any instead of void*.  This seems promising to me, but I am not sure how should we implement the default deleter as C++'s `delete`, like shared_ptr has a default deleter.  

I approve because I guess we need to start programming before we can figure this out.  We can update the document at any future time if necessary.

**Comment by wangkuiyi** [comment]
I tried to make an implementation in https://github.com/PaddlePaddle/Paddle/pull/2587, which works and doesn't rely on boost.

**Comment by qingqing01** [comment]
close this PR, since it had been fixed in https://github.com/PaddlePaddle/Paddle/pull/2587.

**Pull request closed without merging**